tests/estimators/speech_recognition/test_tensorflow_lingvo.py

```diff
@@ -0,0 +1,329 @@
+# MIT License
+#
+# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2020
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+# documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
+# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+# persons to whom the Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+
+import numpy as np
+import pytest
+import tensorflow as tf
+from numpy.testing import assert_allclose, assert_array_equal
+from scipy.io.wavfile import read
+
+from art.estimators.speech_recognition.speech_recognizer import SpeechRecognizerMixin
+from art.estimators.speech_recognition.tensorflow_lingvo import LingvoAsr
+from art.estimators.tensorflow import TensorFlowV2Estimator
+from art.utils import get_file
+
+logger = logging.getLogger(__name__)
+
+
+@pytest.fixture
+def audio_data():
+    """
+    Create audio fixtures of shape (nb_samples=3,) with elements of variable length.
+    """
+    sample_rate = 16000
+    test_input = np.array(
+        [
+            np.zeros(sample_rate),
+            np.ones(sample_rate * 2) * 2e3,
+            np.ones(sample_rate * 3) * 3e3,
+            np.ones(sample_rate * 3) * 3e3,
+        ],
+        dtype=object
+    )
+    return test_input
+
+
+@pytest.fixture
+def audio_batch_padded():
+    """
+    Create audio fixtures of shape (batch_size=2,) with elements of variable length.
+    """
+    sample_rate = 16000
+    frequency_length = (sample_rate // 2 + 1) // 240 * 3
+    test_input = np.zeros((2, sample_rate))
+    test_mask_frequency = np.ones((2, frequency_length, 80))
+    return test_input, test_mask_frequency
+
+
+@pytest.mark.skipif(tf.__version__ != "2.1.0", reason="requires Tensorflow 2.1.0")
+@pytest.mark.skipMlFramework("pytorch", "mxnet", "kerastf", "non_dl_frameworks")
+class TestLingvoAsr:
```

Comment:
The class should be removed and markers placed in front of each test instead. See section 3.1.3 in [wiki doc](https://github.com/Trusted-AI/adversarial-robustness-toolbox/wiki/ART-Unit-Testing) and example class [test_deeplearning_common.py](https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/main/tests/estimators/classification/test_deeplearning_common.py)

Reply:
I understand that we want to bound markers to the method. Therefore, I would suggest to adapt the documentation to explicityly ask for this. But to continue to allowing test classes. Even if we disregard the benefits of using test classes for sharing fixtures or apply marks at the class level, we still allow the benefit to group multiple tests and increase code organization. Also, note that `pytest` will ensure that each test inside a test class has it's own class instance.

What are your opinions @killianlevacher and @beat-buesser ?

Reply:
sure, yes that's a good idea. The most important really is that the markers are all set at the method level and not at the class level. This is to ensure that there is no hierarchical dependency on the markers that regulate which frameworks should be used for each test method. Also as you mentioned, we really need to be careful about not putting setup/teardown methods in these classes then an instead reuse ART fixtures as much as possible. I will update the wiki documentation to reflect this clarification. Thanks for your suggestion.